core/market/decentralized/src/negotiation/common.rs

```diff
@@ -34,36 +38,66 @@ impl CommonBroker {
         subscription_id: &SubscriptionId,
         prev_proposal_id: &ProposalId,
         proposal: &ClientProposal,
+        owner: OwnerType,
     ) -> Result<(Proposal, IsInitial), ProposalError> {
-        // TODO: Everything should happen under transaction.
-        // TODO: Check if subscription is active
-        // TODO: Check if this proposal wasn't already countered.
-        log::debug!("Fetching proposal [{}].", &prev_proposal_id);
+        // Check if subscription is still active.
+        // Note that subscription can be unsubscribed, before we get to saving
+        // Proposal to database. This seems like race conditions, but there's no
+        // danger of data inconsistency. If we won't reject countering Proposal here,
+        // it will be sent to Provider and his counter Proposal will be rejected later.
+        // TODO: We should use validate_subscription function to do this stuff.
+        if owner == OwnerType::Provider {
+            self.store
+                .get_offer(&subscription_id)
+                .await
+                .map_err(|e| match e {
+                    QueryOfferError::Unsubscribed(id) => ProposalError::Unsubscribed(id),
+                    QueryOfferError::Expired(id) => ProposalError::SubscriptionExpired(id),
+                    QueryOfferError::NotFound(id) => ProposalError::NoSubscription(id),
+                    QueryOfferError::Get(..) => {
+                        ProposalError::InternalError(prev_proposal_id.clone(), e.to_string())
+                    }
+                })?;
+        } else {
+            self.store
+                .get_demand(&subscription_id)
+                .await
+                .map_err(|e| match e {
+                    DemandError::NotFound(id) => ProposalError::NoSubscription(id),
+                    _ => ProposalError::InternalError(prev_proposal_id.clone(), e.to_string()),
+                })?;
+        }
```

Comment:
Check if I can get proposal from database and use validate_subscription

Reply:
Moved to: https://github.com/golemfactory/yagna/pull/505
Affects to much code to do this in this PR.
Some tests won't pass until test utils for breaking networking will be merged.